### ISSUE TITLE
Could not establish slave connection to exasol cluster

### ISSUE THREAD

**Issue by 3cham** [EXASOL/spark-exasol-connector#9]
Hi, firstly many thanks for working on this project.  However when I tested with our exasol cluster, I could not perform a simple count() on the dataframe. 

I saw that the schema was inferred correctly:
```
df: org.apache.spark.sql.DataFrame = [COL1: string, COL2: string ... 4 more fields]
```

However the operation using exa-slave connection could not be performed successfully. 
```
com.exasol.jdbc.ConnectFailed: Connection was lost and could not be reestablished.  (SessionID: 1613225809096577815)
	at com.exasol.jdbc.AbstractEXAConnection.reconnect(AbstractEXAConnection.java:3505)
	at com.exasol.jdbc.ServerCommunication.handle(ServerCommunication.java:98)
	at com.exasol.jdbc.AbstractEXAConnection.communication(AbstractEXAConnection.java:2537)
	at com.exasol.jdbc.AbstractEXAConnection.communication_resultset(AbstractEXAConnection.java:2257)
	at com.exasol.jdbc.AbstractEXAStatement.execute(AbstractEXAStatement.java:456)
	at com.exasol.jdbc.EXAStatement.execute(EXAStatement.java:278)
	at com.exasol.jdbc.AbstractEXAStatement.executeQuery(AbstractEXAStatement.java:601)
	at com.exasol.spark.rdd.ExasolRDD.compute(ExasolRDD.scala:123)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:96)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
	at org.apache.spark.scheduler.Task.run(Task.scala:109)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```
Tested using:
- spark 2.3.0
- scala 2.11.8
- ubuntu

I could successfully run a separated test for the exa-slave connection using provided example from exasol (https://www.exasol.com/support/browse/SOL-546). I assume that there shouldn't be any connection problem in between. 

Could you please shed some light on this?

Many thanks.

**Comment by morazow** [comment]
Hello @3cham ,

Thanks for the feedback! 

Yes, this might be one of the known issues. I have had similar problem for instance when doing `.show(10, false)`.

This happens when a Spark `Task` is re-started again. Each task is assigned a single sub connection and once the Task is finished the sub-connection is closed.  

https://github.com/EXASOL/spark-exasol-connector/blob/master/src/main/scala/com/exasol/spark/rdd/ExasolRDD.scala#L116-L118

Could you please confirm from your logs that his happens? 

You can look in Spark executor (or yarn container) logs for this line

https://github.com/EXASOL/spark-exasol-connector/blob/master/src/main/scala/com/exasol/spark/rdd/ExasolRDD.scala#L122

for the second time.

Did this happen on `.count()` or you had some other transformations before count?

Best

**Comment by 3cham** [comment]
Hi @morazow,

yes, it turns out that the main connection was closed too. So could it be a solution when we keep the main connection alive in the driver?

However, my main concern was how could the spark tasks be failed or re-started, when I ran the job locally (not on any cluster), so that there wasnt any preemption happenned.

**Comment by morazow** [comment]
Hello @3cham ,

Have you build the latest commit from repository? Because, the main connection should be recreated again it was closed, [here](https://github.com/EXASOL/spark-exasol-connector/blob/master/src/main/scala/com/exasol/spark/util/ExasolConnectionManager.scala#L72).

Once the Spark application job ends we [close the main connection](https://github.com/EXASOL/spark-exasol-connector/blob/master/src/main/scala/com/exasol/spark/rdd/ExasolRDD.scala#L50-L52), this is also lifetime of SparkContext. However, when running locally Spark re-uses the previous SparkContext, which already closed the main connection. I have faced this situation when running integration tests.  

With above commit this problem should be solved, please let me know if this occurs when running locally using the latest changes.

The `reestablishing of sub connection` error, restarting Spark tasks again, is separate from this. I am going to check this with more Spark actions (`show`, `count`, etc) in cluster environment.

Meanwhile please do not hesitate to let me know if you have suggestions or more feedback.

**Comment by 3cham** [comment]
Hi @morazow , sorry for my late reply. 

Yes, it was tested with the latest commit. count is correctly performed with the docker instance in the integration tests. However, when I try to query the table (mentioned above) in our cluster, the slave connection is closed while the first task is not really completed:

```
18/10/12 14:32:22 INFO ExasolRDD: Connecting to the partition sub connection: jdbc:exa-slave:xxx.xxx.xxx.xxx:20701;slaveID=0;slavetoken=9022XXXXXXXXXXXXX
[Stage 0:>                                                          (0 + 1) / 9]18/10/12 14:34:26 INFO ExasolRDD: Closed a sub connection
```
I don't know how the slave connection is implemented but could it be some kind of network/encryption issue?

**Comment by morazow** [comment]
Hello @3cham,
Could you please also write the query string you are running, or a simplified version of it?

**Comment by 3cham** [comment]
Hi @morazow, could you please confirm that once `connection.EnterParallel()` is called, the query should be simultaneously executed on all the nodes. When I try to execute the query using the slave-connection each after another, I have connection either reset or lost. 

In my case, since it was a local spark context with 1 executor, only one task could be started and it turns out that the query in this task faces the same issue.

**Comment by morazow** [comment]
Hey @3cham,

Yes, that is ideal case. But with sub-connections it does not work like that unfortunately.

I am working on some ideas how to resolve these issue, we can also discuss it next week in Hamburg :)

### First problem

The first problem is once the task is finished we close that particular sub connection. But maybe the Spark context is still running. So next time if that or another task tries to connect previous closed sub connection it fails as closed.

You can reproduce this case with doing collect on a dataframe and then afterwards trying to select or other operation which will fail. For example:
```scala
val df = spark.read.format("exasol")..

df.collect().foreach(println)

df.select("a").count() // this should fail
```

I think this can be easily solved. We can introduce kind of smart connection which actually does not close the connection. Then add a Shutdown hook to close all main / sub connections at the very end.

### Second proble

Second similar issue is kind of similar. Spark start two similar tasks with same sub connection url with same tokens. In this case one of them fails as refused. 

This is a little bit tricky. But one possible solution is to keep internal global state of connections and serialize them into tasks. Then tasks choose the one they need using sub connection url instead of creating a connection.

**Comment by 3cham** [comment]
@morazow: I got what you mean, however the above test is just a standalone program with no spark involved. This program calculates the total number of rows by executing select count(*) on the nodes one after another and adding them together. Please find below the test code:

`   val hosts = conn.GetSlaveHosts()
    val ports = conn.GetSlavePorts()
    System.out.println("using host: " + hosts.mkString(","))
    System.out.println("using port: " + ports.mkString(","))

    val token = conn.GetSlaveToken()

    val subConn = hosts.zip(ports).zipWithIndex.map {
      case ((host, port), idx) => s"jdbc:exa-slave:$host:$port;slavetoken=$token;slaveID=$idx;autocommit=0;encryption=1"
    }

    val numberOfUser = subConn.zipWithIndex.map {
      case (connStr, idx) => {
        System.out.println(s"Querying from instance ID $idx using query: $connStr")
          DriverManager.getConnection(connStr, username, password).asInstanceOf[EXAConnection]
      }}.map {
      case connection => {
        val stmt: Statement = connection.createStatement()
        val res: ResultSet = stmt.executeQuery("select count(*) from TEST.tep")
        val result = res.getInt(1)
        res.close()
        stmt.close()
        connection.close()
        result
      }
    }.sum

    conn.close()

    System.out.println("Total number of users: " + numberOfUser)`

Logs:
`Querying from instance ID 0 using query: jdbc:exa-slave:a.b.c.11:20254;slavetoken=693591696176005203;slaveID=0;autocommit=0;encryption=1
Querying from instance ID 1 using query: jdbc:exa-slave:a.b.c.12:20099;slavetoken=693591696176005203;slaveID=1;autocommit=0;encryption=1
Querying from instance ID 2 using query: jdbc:exa-slave:a.b.c.13:20290;slavetoken=693591696176005203;slaveID=2;autocommit=0;encryption=1
Querying from instance ID 3 using query: jdbc:exa-slave:a.b.c.14:20336;slavetoken=693591696176005203;slaveID=3;autocommit=0;encryption=1
Querying from instance ID 4 using query: jdbc:exa-slave:a.b.c.15:20618;slavetoken=693591696176005203;slaveID=4;autocommit=0;encryption=1
Querying from instance ID 5 using query: jdbc:exa-slave:a.b.c.16:20711;slavetoken=693591696176005203;slaveID=5;autocommit=0;encryption=1
Querying from instance ID 6 using query: jdbc:exa-slave:a.b.c.17:20199;slavetoken=693591696176005203;slaveID=6;autocommit=0;encryption=1
Querying from instance ID 7 using query: jdbc:exa-slave:a.b.c.18:20361;slavetoken=693591696176005203;slaveID=7;autocommit=0;encryption=1
Querying from instance ID 8 using query: jdbc:exa-slave:a.b.c.19:20593;slavetoken=693591696176005203;slaveID=8;autocommit=0;encryption=1
Exception in thread "main" com.exasol.jdbc.ConnectFailed: Connection was lost and could not be reestablished.  (SessionID: 1614670066247435163)
	at com.exasol.jdbc.AbstractEXAConnection.reconnect(AbstractEXAConnection.java:3249)
	at com.exasol.jdbc.ServerCommunication.handle(ServerCommunication.java:85)
	at com.exasol.jdbc.AbstractEXAConnection.communication(AbstractEXAConnection.java:2320)
	at com.exasol.jdbc.AbstractEXAConnection.communication_resultset(AbstractEXAConnection.java:2031)
	at com.exasol.jdbc.AbstractEXAStatement.execute(AbstractEXAStatement.java:456)
	at com.exasol.jdbc.EXAStatement.execute(EXAStatement.java:278)
	at com.exasol.jdbc.AbstractEXAStatement.executeQuery(AbstractEXAStatement.java:601)
	at com.exasol.spark.Main$$anonfun$3.apply(Main.scala:45)
	at com.exasol.spark.Main$$anonfun$3.apply(Main.scala:42)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
	at com.exasol.spark.Main$.main(Main.scala:42)
	at com.exasol.spark.Main.main(Main.scala)`

**Comment by morazow** [comment]
Hey @3cham ,
This good test actually. I do not know the reason yet. I am going to ask here in house.

**Comment by morazow** [comment]
Hey @3cham,
I have talked with Thomas here at Exasol. 

This will not work because, before fetching any data from a sub-connection, all sub-connections should already have query executed or prepared.

Could you please try creating and exsuting statement in first map after creating a connection. Then in second map to try to get data from result set?

Additionally, you can execute the query in main connection and use the handle in sub-connections.

**Comment by 3cham** [comment]
Since we all know that it depends on sub connection implementation, I will close this issue :)
